libp2p/protocols/kademlia/kademlia.nim

```diff
@@ -0,0 +1,521 @@
+      # This is for test, but we really want to call iterative here
+      # Can we just respond straight away here?
+      # TODO: Wrong, should have value
+      # TODO: HERE ATM, let's iterate over KadPeers and add RPC Peer construct
+      # Kadpeer -> RPC Peer
+      var peers: seq[Peer]
+      for kp in res:
+        debug "rpc peer", id = kp.peerInfo.peerId, addrs = kp.peerInfo.addrs
+
+        # TODO: Get bytes encoding of addrs in protobuf
+        var mas: seq[seq[byte]]
+        for ma in kp.peerInfo.addrs:
+          # FIXME: Broken
+          mas.add(cast[seq[byte]]($ma))
+        var p = Peer(id: kp.peerInfo.peerId.getBytes(),
+                     addrs: mas)
+
+        peers.add(p)
+        debug "closerPeers", peers = peers
+
+      var closerPeers = CloserPeers(peers: peers)
+      var messageType = messages.MessageType(strtype: "FIND_NODE_REPLY")
+      let replyMsg = RPCMSg(messageType: messageType,
+                            key: m.key,
+                            closerPeers: some(closerPeers))
+      debug "find node reply rpc", msg = replyMsg
+      var msgs: seq[RPCMSg]
+      msgs.add(replyMsg)
+      await peer.send(msgs)
+      await p.findNodeHandler(res)
+    elif m.messageType.strtype == "FIND_NODE_REPLY":
+      # TODO: We want to call findNodeHandler here
+      # Need to convert KadPeer, Peer etc
+      debug "FIND_NODE_REPLY", m = m, key = m.key, closerPeer = m.closerPeers
+
+method handleConn*(p: KadProto,
+                   conn: Connection,
+                   proto: string) {.base, async.} =
+  trace "handleConn remote connection", proto = proto
+
+  # TODO: see pubsub/handleConn
+  #await conn.writeLp("Hello!")
+  #await conn.close()
+  if isNil(conn.peerInfo):
+    trace "no valid PeerId for peer"
+    await conn.close()
+    return
+
+  proc handler(peer: KadPeer, msgs: seq[RPCMsg]) {.async.} =
+    discard p.rpcHandler(peer, msgs)
+
+  let peer = p.getPeer(conn.peerInfo, proto)
+  peer.handler = handler
+
+  await peer.handle(conn) # spawn peer read loop
+  # TODO: Handle cleanup, etc
+
+method init(p: KadProto) =
+  #{.base, gcsafe, async.} =
+  # handle incoming connections in closure
+  # TODO: First hit this, then generalize a la pubsub/handleConn
+  # Whenever we get a connection this should be triggered
+  proc handle(conn: Connection, proto: string) {.async, gcsafe.} =
+    # main handler that gets triggered on connection for protocol string
+    # triggered upon connection for protocol string
+    await p.handleConn(conn, proto)
+
+  p.handler = handle # set proto handler
+  p.codec = KadCodec # init proto with the correct string id
+
+method start*(p: KadProto) {.async, base.} =
+  # start kad
+  discard
+
+method stop*(p: KadProto) {.async, base.} =
+  # stop kad
+  discard
+
+# When we add contact we should also getPeer or set their peer info
+# XXX: Do we want to listen to them as well?
+method addContact*(p: KadProto, contact: PeerInfo) {.base, gcsafe.} =
```

Comment:
Is the intention here for these Kademlia-specific operations to also be defined as methods? As far as I can tell, you don't intend to introduce types derived from `KadProto` and these could be regular procs.